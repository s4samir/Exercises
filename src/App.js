/*
 * Authored by Samir Kishore
 */
import React, { useState } from "react";
import "./App.scss";
import MineField from "./Components/MineField/MineField";
import Controls from "./Components/Controls/Controls";
import TopForm from "./Components/TopForm/TopForm";
import ResultDialog from "./Components/ResultDialog/ResultDialog";

function App() {
  const [mineMap, setMineMap] = useState(null);
  const [level, setLevel] = useState([5, 5]);
  const [isHome, setIsHome] = useState(true);
  const [timer, setTimer] = useState(false);
  const [result, setResult] = useState("");
  const onLevelChange = (value) => {
    const [row, col] = value;
    const mine = Math.floor(0.4 * (parseInt(row) * parseInt(col)));
    setLevel([+row, +col]);
    setMineMap(getUniqueRandom(+row, +col, mine));
    setIsHome(false);
  };

  const gameView = () => (
    <>
      <Controls startTheTimer={timer} onNewGame={reset} />

      <div className="minegame">
        <MineField
          row={level[0]}
          col={level[1]}
          mineMap={mineMap}
          startTimer={(val) => setTimer(val)}
          closeGame={handleGameResult}
        />
      </div>
      <ResultDialog onClose={reset} result={result} />
    </>
  );

  const handleGameResult = (result) => {
    setResult(result);
  };

  const reset = () => {
    setResult("");
    setIsHome(true);
  };

  const homeView = () => <TopForm onLevelChange={onLevelChange} />;

  return <div className="App">{isHome ? homeView() : gameView()} </div>;
}

const getUniqueRandom = (row, col, mines) => {
  const nums = new Set();
  while (nums.size !== mines) {
    const x = Math.floor(Math.random() * row) + 1,
      y = Math.floor(Math.random() * col) + 1;
    nums.add(`${x}${y}`);
  }
  return nums;
};

export default App;
