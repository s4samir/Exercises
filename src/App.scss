.App {
  text-align: center;
  font-size: 20px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
}
.minegame{
  display: inline-block;
  margin: auto;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
