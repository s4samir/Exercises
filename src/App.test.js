/*
 * Authored by Samir Kishore
 */
import React from "react";
import App from "./App";
import { create } from "react-test-renderer";

test("renders learn react link", () => {
  const wrapper = create(<App />);
  expect(wrapper.toJSON()).toMatchSnapshot();
});
