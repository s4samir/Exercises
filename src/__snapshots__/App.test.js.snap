// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  className="App"
>
  <div
    className="TopForm"
  >
    <div>
      <h2
        className="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom"
      >
        Mine Sweeper
      </h2>
    </div>
    <div>
      <div
        aria-label="large outlined primary button group"
        className="MuiButtonGroup-root"
        role="group"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Beginner
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Intermediate
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Advanced
          </span>
        </button>
      </div>
    </div>
  </div>
   
</div>
`;
